sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -5056,128 +4950,45 @@ case class ArrayCompact(child: Expression)
-        TypeCoercion.findTightestCommonType(e1, e2) match {
-          case Some(dt) => Seq(ArrayType(dt, hasNull), dt)
-          case _ => Seq.empty
-        }
-      case _ => Seq.empty
-    }
-  }
+  override def prettyName: String = "array_append"
 
-  override def checkInputDataTypes(): TypeCheckResult = {
-    (left.dataType, right.dataType) match {
-      case (ArrayType(e1, _), e2) if DataTypeUtils.sameType(e1, e2) => TypeCheckResult.TypeCheckSuccess
-      case (ArrayType(e1, _), e2) => DataTypeMismatch(
-        errorSubClass = "ARRAY_FUNCTION_DIFF_TYPES",
-        messageParameters = Map(
-          "functionName" -> toSQLId(prettyName),
-          "leftType" -> toSQLType(left.dataType),
-          "rightType" -> toSQLType(right.dataType),
-          "dataType" -> toSQLType(ArrayType)
-        ))
-      case _ =>
-        DataTypeMismatch(
-          errorSubClass = "UNEXPECTED_INPUT_TYPE",
-          messageParameters = Map(
-            "paramIndex" -> "0",
-            "requiredType" -> toSQLType(ArrayType),
-            "inputSql" -> toSQLExpr(left),
-            "inputType" -> toSQLType(left.dataType)
-          )
-        )
-    }
+  override protected def withNewChildrenInternal(
+      newLeft: Expression, newRight: Expression): ArrayAppend = {
+    copy(left = newLeft, right = newRight)
   }
+}
 
-  override def eval(input: InternalRow): Any = {
-    val value1 = left.eval(input)
-    if (value1 == null) {
-      null
-    } else {
-      val value2 = right.eval(input)
-      nullSafeEval(value1, value2)
-    }
-  }
+@ExpressionDescription(
+  usage = """
+      _FUNC_(array, element) - Add the element at the beginning of the array passed as first
+      argument. Type of element should be the same as the type of the elements of the array.
+      Null element is also prepended to the array. But if the array passed is NULL
+      output is NULL
+    """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array('b', 'd', 'c', 'a'), 'd');
+       ["d","b","d","c","a"]
+      > SELECT _FUNC_(array(1, 2, 3, null), null);
+       [null,1,2,3,null]
+      > SELECT _FUNC_(CAST(null as Array<Int>), 2);
+       NULL
+  """,
+  group = "array_funcs",
+  since = "3.5.0")
+case class ArrayPrepend(left: Expression, right: Expression) extends InsertArrayOneSide {
 
-  override protected def nullSafeEval(arr: Any, elementData: Any): Any = {
-    val arrayData = arr.asInstanceOf[ArrayData]
-    val numberOfElements = arrayData.numElements() + 1
-    if (numberOfElements > ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH) {
-      throw QueryExecutionErrors.concatArraysWithElementsExceedLimitError(numberOfElements)
-    }
-    val finalData = new Array[Any](numberOfElements)
-    arrayData.foreach(elementType, finalData.update)
-    finalData.update(arrayData.numElements(), elementData)
-    new GenericArrayData(finalData)
-  }
+  override lazy val replacement: Expression = ArrayInsert(left, Literal(0), right)
```

Comment:
not related to this PR, but isn't the implementation of `array_insert` wrong? The doc says: `Places val into index pos of array x. Array indices start at 1, or start from the end if index is negative.`

According to the doc, `array_insert(arr, 1, val)` should put `val` as the first element in this array. `array_insert(arr, 0, val)` should fail as 0 is not a valid index. However, the current implementation of `array_insert` uses 0-based index. cc @zhengruifeng @HyukjinKwon @dtenedor

Reply:
@cloud-fan it is still 1-based, but also treat 0 as the first item. see the discussion https://github.com/apache/spark/pull/38867#discussion_r1093243515

```
+----------------------------------+----------------------------------+----------------------------------+
|array_insert(array(1, 2, 3), 1, 4)|array_insert(array(1, 2, 3), 0, 4)|array_insert(array(1, 2, 3), 2, 4)|
+----------------------------------+----------------------------------+----------------------------------+
|                      [4, 1, 2, 3]|                      [4, 1, 2, 3]|                      [1, 4, 2, 3]|
+----------------------------------+----------------------------------+----------------------------------+
```

Reply:
Sorry for missing that discussion. I think failing is more consistent with other functions that use 1-based index.

And here we should use `1` as the index to be more future-proof.

Reply:
got it, let me send a fix for it.

Reply:
If we do want to reserve index `0` for some special behavior in `array_insert`, I think appending the value to the end of the array seems more convenient?

Reply:
I chose to remove support for index `0` ...

Reply:
I think we don't have to reserve index `0`, make it fails seems good to me

Reply:
great, let's fix it soon before the 3.4 release. also cc @xinrong-meng